leaf/src/proxy/obfs/http.rs

```diff
@@ -0,0 +1,211 @@
+                        ready!(tokio_util::io::poll_read_buf(Pin::new(stream), cx, res_buf))?;
+                    if read_len == 0 {
+                        return Poll::Ready(Err(io::Error::new(
+                            io::ErrorKind::Other,
+                            "obfs response too short",
+                        )));
+                    }
+                    let req_body_pos = match memmem::find(res_buf, b"\r\n\r\n") {
+                        Some(p) => p + 4,
+                        None => continue,
+                    };
+                    let mut payload = Cursor::new(std::mem::take(res_buf));
+                    payload.set_position(req_body_pos as u64);
+                    *read_state = ReadState::ConsumingResponse { res: payload };
+                }
+                ReadState::ConsumingResponse { res } => {
+                    let remaining = &res.get_ref()[res.position() as usize..];
+                    let to_write = remaining.len().min(buf.remaining());
+                    buf.put_slice(&remaining[..to_write]);
+                    res.set_position(res.position() + to_write as u64);
+                    if res.position() as usize == res.get_ref().len() {
+                        *read_state = ReadState::Transfer;
+                    }
+                    return Poll::Ready(Ok(()));
+                }
+                ReadState::Transfer => return Pin::new(stream).poll_read(cx, buf),
+            };
+        }
+    }
+}
+
+impl AsyncWrite for Stream {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let Self {
+            write_state,
+            stream,
+            ..
+        } = &mut *self;
+        loop {
+            match write_state {
+                WriteState::Initial { req_line } => {
+                    let req = generate_http_request(req_line, buf);
+                    *write_state = WriteState::WritingRequest(Cursor::new(req));
+                }
+                WriteState::WritingRequest(req) => {
+                    ready!(poll_write_buf(Pin::new(stream), cx, req))?;
+                    if req.position() as usize == req.get_ref().len() {
+                        *write_state = WriteState::Transfer;
+                        return Poll::Ready(Ok(buf.len()));
+                    }
+                }
+                WriteState::Transfer => break,
+            }
+        }
+        Pin::new(&mut *stream).poll_write(cx, buf)
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let Self {
+            write_state,
+            stream,
+            ..
+        } = &mut *self;
+        loop {
+            match write_state {
+                WriteState::Initial { .. } => return Poll::Ready(Ok(())),
+                WriteState::WritingRequest(req) => {
+                    ready!(poll_write_buf(Pin::new(stream), cx, req))?;
```

Comment:
This doesn't seem to be a correct handling. If the underlying buffer is already full (e.g. there is a very large initial data being sent with the request), this will return Pending, the IO object is never flushed.

Reply:
If its internal buffer is full, it is the underlying stream’s responsibility to flush and make room for new data. A flush call is just to ask the stream to send out whatever in its buffer, and is not required upon normal write calls. Also see [tokio::io::BufWriter::poll_write](https://docs.rs/tokio/1.27.0/src/tokio/io/util/buf_writer.rs.html#119-134)

You can think of it another way. Given the buffer is full, if `poll_write` returns `Pending`, it would have at least registered the waker to wake it up at some time for us to continue writing, which means there is some IO under the hood. It should not do nothing.

Reply:
Why do you write the req here while it's already done in poll_write?

Reply:
Fixed in https://github.com/eycorsican/leaf/pull/366/commits/7b8620b17b036d74db582910f6738b01efadb462 .